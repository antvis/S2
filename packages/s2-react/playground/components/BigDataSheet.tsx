import { SpreadSheet, type S2DataConfig } from '@antv/s2';
import React from 'react';

import {
  SheetComponent,
  type SheetComponentOptions,
  type SheetComponentsProps,
} from '../../src';
import { usePlaygroundContext } from '../context/playground.context';

const s2Options: SheetComponentOptions = {
  width: 1600,
  height: 480,
  interaction: {
    scrollSpeedRatio: {
      vertical: 10,
      horizontal: 10,
    },
  },
};

export function generateRawData(
  row: Record<string, number>,
  col: Record<string, number>,
) {
  const res: Record<string, any>[] = [];
  const rowKeys = Object.keys(row);
  const colKeys = Object.keys(col);

  for (let i = 0; i < row[rowKeys[0]]; i++) {
    for (let j = 0; j < row[rowKeys[1]]; j++) {
      for (let m = 0; m < col[colKeys[0]]; m++) {
        for (let n = 0; n < col[colKeys[1]]; n++) {
          res.push({
            province: `p${i}`,
            city: `c${j}`,
            type: `type${m}`,
            subType: `subType${n}`,
            number: i * n,
          });
        }
      }
    }
  }

  return res;
}

const s2DataConfig: S2DataConfig = {
  fields: {
    rows: ['type', 'subType'],
    columns: ['province', 'city'],
    values: ['number'],
  },
  data: generateRawData(
    { province: 10, city: 100 },
    { type: 10, subType: 100 },
  ),
};

type CustomGridProps = Partial<SheetComponentsProps>;
const sheetType = 'pivot';

export const BigDataSheet = React.forwardRef<SpreadSheet, CustomGridProps>(
  (props, ref) => {
    const context = usePlaygroundContext();

    return (
      <SheetComponent
        {...props}
        {...context}
        sheetType={sheetType}
        dataCfg={s2DataConfig}
        options={s2Options}
        ref={ref}
      />
    );
  },
);
