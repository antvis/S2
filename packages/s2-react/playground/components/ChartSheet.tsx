/* eslint-disable import/no-extraneous-dependencies */
import { renderToMountedElement, stdlib } from '@antv/g2';
import { SpreadSheet, type ThemeCfg } from '@antv/s2';
import React from 'react';
import { ChartDataConfig } from '../../__tests__/data/data-g2-chart';
import {
  SheetComponent,
  type SheetComponentOptions,
  type SheetComponentsProps,
} from '../../src/components';
import { usePlaygroundContext } from '../context/playground.context';

const options: SheetComponentOptions = {
  height: 900,
  interaction: {
    copy: { enable: true, withFormat: true },
    brushSelection: {
      rowCell: true,
      colCell: true,
      dataCell: false,
    },
    selectedCellMove: true,
    selectedCellHighlight: true,
    selectedCellsSpotlight: true,
    hoverFocus: true,
    hoverHighlight: false,
    multiSelection: true,
    overscrollBehavior: 'none',
  },
  style: {},
};

const theme: ThemeCfg['theme'] = {
  dataCell: {},
};

const onDataCellRender: SheetComponentsProps['onDataCellRender'] = (cell) => {
  // 普通数值单元格正常展示
  if (!cell.isChartData()) {
    return;
  }

  const chartOptions = cell.getRenderChartOptions();

  // https://g2.antv.antgroup.com/manual/extra-topics/bundle#g2stdlib
  renderToMountedElement(chartOptions, {
    group: cell,
    library: stdlib(),
  });
};

export const ChartSheet: React.FC<
  Partial<SheetComponentsProps> & React.RefAttributes<SpreadSheet>
> = React.forwardRef((props, ref) => {
  const context = usePlaygroundContext();

  return (
    <SheetComponent
      {...props}
      {...context}
      sheetType="chart"
      themeCfg={{ theme }}
      dataCfg={ChartDataConfig}
      options={options}
      ref={ref}
      adaptive={true}
      onDataCellRender={onDataCellRender}
      header={{
        title: 'S2 & G2',
        description: '单元格内绘制 G2 图表',
        export: { open: true },
      }}
    />
  );
});
