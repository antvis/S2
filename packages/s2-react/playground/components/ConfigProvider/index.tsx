import { getLang, type ThemeName } from '@antv/s2';
import { ConfigProvider as AntdConfigProvider, theme } from 'antd';
import enUS from 'antd/es/locale/en_US';
import ruRU from 'antd/es/locale/ru_RU';
import zhCN from 'antd/es/locale/zh_CN';
import React from 'react';

export interface ConfigProviderProps {
  children: React.ReactNode;
  themeName?: ThemeName;
}

export const ConfigProvider: React.FC<ConfigProviderProps> = (props) => {
  const { children, themeName } = props;
  const lang = getLang();
  // eslint-disable-next-line no-nested-ternary
  const locale = lang === 'zh_CN' ? zhCN : lang === 'ru_RU' ? ruRU : enUS;
  const isDarkTheme = themeName === 'dark';

  return (
    <AntdConfigProvider
      locale={locale}
      theme={{
        algorithm: isDarkTheme ? theme.darkAlgorithm : theme.defaultAlgorithm,
      }}
    >
      {children}
    </AntdConfigProvider>
  );
};
