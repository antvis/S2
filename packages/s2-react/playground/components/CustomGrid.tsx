/* eslint-disable no-console */
import type { S2DataConfig, SpreadSheet, ThemeCfg } from '@antv/s2';
import {
  customColGridFields,
  customRowGridFields,
} from '@antv/s2/__tests__/data/custom-grid-fields';
import { CustomGridData } from '@antv/s2/__tests__/data/data-custom-grid';
import { Radio, Space, Switch } from 'antd';
import React from 'react';
import {
  SheetComponent,
  type SheetComponentOptions,
  type SheetComponentsProps,
} from '../../src';
import { meta } from '../../__tests__/data/mock-dataset.json';
import { onSheetMounted } from '../utils';
import { usePlaygroundContext } from '../context/playground.context';
import { ResizeConfig } from './ResizeConfig';

export const customRowGridOptions: SheetComponentOptions = {
  width: 1000,
  height: 480,
  hierarchyType: 'grid',
  cornerText: '自定义角头标题',
};

/**
 * 平铺模式-自定义行头
 */
export const pivotSheetCustomRowGridDataCfg: S2DataConfig = {
  data: CustomGridData,
  meta: [
    ...meta,
    {
      field: 'a-1',
      name: '层级1',
    },
    {
      field: 'a-1-1',
      name: '层级2',
    },
    {
      field: 'measure-1',
      name: '层级3',
    },
    {
      field: 'measure-1',
      formatter: (value) => `#-${value}`,
    },
  ],
  fields: customRowGridFields,
};

/**
 * 平铺模式-自定义列头
 */
export const pivotSheetCustomColGridDataCfg: S2DataConfig = {
  data: CustomGridData,
  meta: [
    ...meta,
    {
      field: 'a-1',
      name: '指标1',
      formatter: (value) => `#-${value}`,
    },
    {
      field: 'a-1-1',
      name: '层级2',
    },
    {
      field: 'measure-1',
      formatter: (value) => `#-${value}`,
    },
  ],
  fields: customColGridFields,
};

enum CustomType {
  Row = 'row',
  Col = 'col',
  All = 'all',
}

type CustomGridProps = Partial<SheetComponentsProps>;

export const CustomGrid = React.forwardRef<SpreadSheet, CustomGridProps>(
  (props, ref) => {
    const context = usePlaygroundContext();
    const { logHandler } = context;
    const [customType, setCustomType] = React.useState<CustomType>(
      (localStorage.getItem('debugCustomType') as unknown as CustomType) ||
        CustomType.Row,
    );
    const [options, setOptions] = React.useState<SheetComponentOptions>({
      ...customRowGridOptions,
      hierarchyType: 'grid',
      interaction: {
        overscrollBehavior: 'none',
      },
    });
    const [themeCfg, setThemeCfg] = React.useState<ThemeCfg>({
      name: 'default',
      ...context.themeCfg,
    });
    const [sheetType, setSheetType] =
      React.useState<SheetComponentsProps['sheetType']>('pivot');

    const dataCfg =
      customType === CustomType.Row
        ? pivotSheetCustomRowGridDataCfg
        : pivotSheetCustomColGridDataCfg;

    return (
      <>
        <Space style={{ marginBottom: 20 }}>
          <Radio.Group
            value={customType}
            onChange={(e) => {
              setSheetType('pivot');
              setCustomType(e.target.value);
            }}
          >
            <Radio.Button value={CustomType.Row}>自定义行头</Radio.Button>
            <Radio.Button value={CustomType.Col}>自定义列头</Radio.Button>
            <Radio.Button value={CustomType.All} disabled>
              TODO: 自定义行头和列头
            </Radio.Button>
          </Radio.Group>
          <Switch
            checkedChildren="树状模式"
            unCheckedChildren="平铺模式"
            checked={options.hierarchyType === 'tree'}
            disabled={sheetType === 'table'}
            onChange={(checked) => {
              setOptions({
                hierarchyType: checked ? 'tree' : 'grid',
              });
            }}
          />
          <Switch
            checkedChildren="序号开"
            unCheckedChildren="序号关"
            checked={options.seriesNumber?.enable}
            onChange={(checked) => {
              setOptions({
                seriesNumber: {
                  enable: checked,
                },
              });
            }}
          />
          <Switch
            checkedChildren="透视表"
            unCheckedChildren="明细表"
            checked={sheetType === 'pivot'}
            disabled={customType !== CustomType.Col}
            onChange={(checked) => {
              setSheetType(checked ? 'pivot' : 'table');
            }}
          />
        </Space>
        <Space style={{ marginBottom: 20, display: 'flex' }}>
          <ResizeConfig
            options={options}
            setOptions={setOptions}
            setThemeCfg={setThemeCfg}
          />
        </Space>

        <SheetComponent
          {...props}
          {...context}
          sheetType={sheetType}
          dataCfg={dataCfg}
          options={options}
          themeCfg={themeCfg}
          ref={ref}
          onMounted={onSheetMounted}
          onLayoutResize={logHandler('onLayoutResize')}
        />
      </>
    );
  },
);
