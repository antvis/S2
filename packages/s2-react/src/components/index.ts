export { BaseSheet } from './sheets/base-sheet';
export { TableSheet } from './sheets/table-sheet';
export { GridAnalysisSheet } from './sheets/grid-analysis-sheet';
export { StrategySheet } from './sheets/strategy-sheet';
export { AdvancedSort } from './advanced-sort';
export { DrillDown } from './drill-down';
export { TooltipComponent } from './tooltip';
export { Switcher } from './switcher';
export * from './sheets';
export * from './sheets/interface';
