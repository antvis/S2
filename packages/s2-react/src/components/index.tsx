import React from 'react';
import debounceRender from 'react-debounce-render';
import { BaseSheet } from './sheets/base-sheet';
import { TabularSheet } from './sheets/tabular-sheet';
import { TableSheet } from './sheets/table-sheet';
import { SpreadsheetProps } from './sheets/interface';
export { DrillDown } from './drill-down';
export { Switcher } from './switcher';
export { AdvancedSort } from './advanced-sort';
export * from './sheets';

export const SheetComponent: React.FC<SpreadsheetProps> = debounceRender(
  (props: SpreadsheetProps) => {
    const { sheetType } = props;
    switch (sheetType) {
      case 'table':
        return <TableSheet {...props} />;
      case 'tabular':
        return <TabularSheet {...props} />;
      default:
        return <BaseSheet {...props} />;
    }
  },
  100,
);
