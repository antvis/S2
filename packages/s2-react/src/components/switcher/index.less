@prefix: antv-s2-switcher;

.@{prefix}-entry-button {
  display: inline-flex !important;
  align-items: center;
}
