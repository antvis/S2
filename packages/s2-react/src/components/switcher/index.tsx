import { Button, Popover, type PopoverProps } from 'antd';
import React, {
  type FC,
  type ReactNode,
  useState,
  isValidElement,
} from 'react';
import { i18n } from '@antv/s2';
import cls from 'classnames';
import { SwitcherIcon } from '../icons';
import { SwitcherContent, type SwitcherContentProps } from './content';
import { getSwitcherClassName } from './util';
import './index.less';

export interface SwitcherProps
  extends Omit<SwitcherContentProps, 'onToggleVisible'> {
  title?: ReactNode;
  // ref: https://ant.design/components/popover-cn/#API
  popover?: PopoverProps;
  disabled?: boolean;
}

export const Switcher: FC<SwitcherProps> = React.memo(
  ({ title, popover, disabled, ...otherProps }) => {
    const [visible, setVisible] = useState(false);
    const onToggleVisible = () => {
      setVisible((prev) => !prev);
    };

    return (
      <Popover
        open={!disabled && visible}
        content={
          <SwitcherContent {...otherProps} onToggleVisible={onToggleVisible} />
        }
        onOpenChange={onToggleVisible}
        trigger="click"
        placement="bottomLeft"
        destroyTooltipOnHide
        {...popover}
        overlayClassName={cls(
          getSwitcherClassName('switcher-overlay'),
          popover?.overlayClassName,
        )}
      >
        {isValidElement(title) ? (
          title
        ) : (
          <Button
            className={getSwitcherClassName('entry-button')}
            size="small"
            disabled={disabled}
            icon={<SwitcherIcon />}
          >
            {title || i18n('行列切换')}
          </Button>
        )}
      </Popover>
    );
  },
);

Switcher.displayName = 'Switcher';
