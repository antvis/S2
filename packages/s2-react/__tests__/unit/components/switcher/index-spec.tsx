import React from 'react';
import { render } from '@testing-library/react';
import { Switcher } from '@/components/switcher';

describe('switcher component test', () => {
  test('should render component', () => {
    const result = render(<Switcher title="title" />);

    expect(result.asFragment()).toMatchSnapshot();
  });
});
