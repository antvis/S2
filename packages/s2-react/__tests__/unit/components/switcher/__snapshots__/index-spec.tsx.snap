// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`switcher component test should render component 1`] = `
<DocumentFragment>
  <button
    class="ant-btn css-dev-only-do-not-override-p7e5j5 ant-btn-default ant-btn-sm antv-s2-switcher-entry-button"
    type="button"
  >
    <span
      class="ant-btn-icon"
    >
      <div />
    </span>
    <span>
      title
    </span>
  </button>
</DocumentFragment>
`;
